kubelet & kube-proxy: ignore sysctl errors and rlimit errors when running in UserNS (for rootless)

AkihiroSuda opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature


**What this PR does / why we need it**:
Ignore sysctl errors and rlimit errors when running in UserNS.

These changes are required for Rootless mode. See [KEP](https://github.com/kubernetes/enhancements/pull/1371) for the further information.

**Special notes for your reviewer**:

How to test rootless mode with this PR:
* Run `make kubelet kube-proxy`
* Git clone  https://github.com/rootless-containers/usernetes and run `make`
* Replace the `kubelet` and `kube-proxy` binaries inside the` bin` directory of the Usernetes  to test the codes in this PR. To just try rootless mode, no need to replace the binaries.
* Configure the host to use cgroup v2: https://github.com/rootless-containers/usernetes#enable-cgroup-v2
* Run the Usernetes installer: `./install.sh`
* `export KUBECONFIG="$HOME/.config/usernetes/master/admin-localhost.kubeconfig"`
* Run `kubectl`

To run the smoke test suite, run `./hack/smoketest-binaries.sh` instead of `./install.sh ...`.

After this PR gets merged, I'm planning to add kind tests with Rootless Docker (or Podman).


**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Adds feature gate `KubeletInUserNamespace` which enables support for running kubelet in a user namespace.

The user namespace has to be created before running kubelet.
All the node components such as CRI need to be running in the same user namespace.

When the feature gate is enabled, kubelet ignores errors that happens during setting the following sysctl values: `vm.overcommit_memory`, `vm.panic_on_oom`, `kernel.panic`, `kernel.panic_on_oops`, `kernel.keys.root_maxkeys`, `kernel.keys.root_maxbytes`. (These sysctl values for the host, not for the containers)

kubelet also ignores an error during opening `/dev/kmsg`.
This feature gate also allows kube-proxy to ignore an error during setting `RLIMIT_NOFILE`.

This feature gate is especially useful for running Kubernetes inside Rootless Docker/Podman with `kind` or `minikube`.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2033-kubelet-in-userns-aka-rootless
```

- - -

The patches are from Usernetes project: https://github.com/rootless-containers/usernetes/tree/master/src/patches/kubernetes

KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2033-kubelet-in-userns-aka-rootless
Tracking issue: https://github.com/kubernetes/enhancements/issues/2033
Documentation PR: https://github.com/kubernetes/website/pull/28827

### kubelet changes
Ignore errors that happens during setting the following sysctl values:
- `vm.overcommit_memory`
- `vm.panic_on_oom`
- `kernel.panic`
- `kernel.panic_on_oops`
- `kernel.keys.root_maxkeys`
- `kernel.keys.root_maxbytes`

Also ignores an error during opening `/dev/kmsg`.

Gated by `KubeletInUserNamespace` feature gate.

### kube-proxy changes
Ignore an error during setting `RLIMIT_NOFILE`

AkihiroSuda committed:
kubelet/cm: ignore sysctl error when running in userns

Errors during setting the following sysctl values are ignored:
- vm.overcommit_memory
- vm.panic_on_oom
- kernel.panic
- kernel.panic_on_oops
- kernel.keys.root_maxkeys
- kernel.keys.root_maxbytes

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
kube-proxy: allow running in userns

Ignore an error during setting RLIMIT_NOFILE.

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda commented:
cc @giuseppe @cyphar @rhatdan @kentatada FYI

hakman commented:
/ok-to-test

tedyu commented on pkg/kubelet/cm/cgroup_manager_linux.go:
```diff
@@ -155,7 +159,10 @@ func (l *libcontainerAdapter) newManager(cgroups *libcontainerconfigs.Cgroup, pa
-			return cgroupsystemd.NewUnifiedManager(cgroups, paths["memory"], false), nil
+			return cgroupsystemd.NewUnifiedManager(cgroups, paths["memory"], l.rootless), nil
+		}
+		if l.rootless {
```
Should kubefeatures.Rootless be checked here ?

AkihiroSuda replied:
The feature gate is checked in `NewCgroupManager()`: https://github.com/kubernetes/kubernetes/pull/92863/files#diff-fa76bb6ae2d9b4bb3d023737fe5e6029R204

tedyu replied:
thanks

AkihiroSuda commented:
/retest

AkihiroSuda commented:
> This PR [may require API review](https://git.k8s.io/community/sig-architecture/api-review-process.md#what-apis-need-to-be-reviewed).
> 
> If so, when the changes are ready, [complete the pre-review checklist and request an API review](https://git.k8s.io/community/sig-architecture/api-review-process.md#mechanics).
> 
> Status of requested reviews is tracked in the [API Review project](https://github.com/orgs/kubernetes/projects/13).

This PR doesn't change any API other than `KubeletConfiguration`, which seems out of [the scope of "API change"](https://github.com/kubernetes/community/blob/4e7a77958f331abeb3bb03729a0e41a8fdd9c02b/sig-architecture/api-review-process.md#what-parts-of-a-pr-are-api-changes)

AkihiroSuda commented:
CI failure doesn't seem related

giuseppe commented on pkg/kubelet/cm/cgroup_manager_linux.go:
```diff
@@ -133,18 +133,22 @@ func IsSystemdStyleName(name string) bool {
 func (l *libcontainerAdapter) newManager(cgroups *libcontainerconfigs.Cgroup, paths map[string]string) (libcontainercgroups.Manager, error) {
 	switch l.cgroupManagerType {
 	case libcontainerCgroupfs:
+		if l.rootless {
```
do we want to address using cgroupfs and running in a new cgroup namespace?

AkihiroSuda replied:
Could you be more specific? I think we are going to support only systemd (as in Podman)

giuseppe replied:
I mean we can create a separate environment with cgroup fully owned by the rootless user, where we run both kubernetes and the container runtime.
Similar to:
```
$ systemd-run --user --scope unshare -rCm bash
# mount -t cgroup2 cgroup2 /tmp && mount --move /tmp /sys/fs/cgroup
```
Now /sys/fs/cgroup is fully owned by the rootless user and we can use the cgroupfs backend in the same way as root.

Wouldn't that work?

AkihiroSuda replied:
```console
$ cat /etc/systemd/system/user@.service.d/delegate-cgroup.conf 
[Service]
Delegate=pids memory cpu cpuset io
$ cat /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/cgroup.controllers 
cpuset cpu io memory pids
$ systemd-run --user --scope unshare -rCm bash
Running scope as unit: run-r31deaab786d54940adb0d0d9e7565247.scope
#  mount -t cgroup2 cgroup2 /tmp && mount --move /tmp /sys/fs/cgroup
mount: /sys/fs/cgroup: filesystem was mounted, but failed to update userspace mount table.
# cat /sys/fs/cgroup/cgroup.controllers 
memory pids
```

Controllers are not delegated as expected

AkihiroSuda replied:
@giuseppe 

Can't write to subtree_control
```console
$ systemd-run --user --scope -p "Delegate=cpuset cpu io memory pids" unshare -rCm sh -c "mount -t cgroup2 cgroup2 /tmp && mount -n  --move /tmp /sys/fs/cgroup && cd /sys/fs/cgroup && bash"
Running scope as unit: run-rc35e9a40b0f64c76aa38bc81ce83eb43.scope
# cat cgroup.controllers 
cpuset cpu io memory pids
# cat cgroup.type 
domain
# echo +memory > cgroup.subtree_control 
bash: echo: write error: Device or resource busy
```

AkihiroSuda replied:
cgroupns doesn't work expected when containerd and rootlesskit are executed as separate services and `/sys/fs/cgroup` is mounted with `nsdelegate`: https://github.com/rootless-containers/usernetes/pull/178#issuecomment-656039720

SergeyKanzhelev commented on test/e2e_node/node_container_manager_test.go:
```diff
@@ -167,7 +167,10 @@ func runTest(f *framework.Framework) error {
-	cgroupManager := cm.NewCgroupManager(subsystems, oldCfg.CgroupDriver)
+	cgroupManager, err := cm.NewCgroupManager(subsystems, oldCfg.CgroupDriver, false)
+	if err != nil {
+		return nil
```
is it intentionally returns `nil` here?

AkihiroSuda replied:
no, fixed

AkihiroSuda commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -209,7 +210,10 @@ func NewCustomProxier(loadBalancer LoadBalancer, listenIP net.IP, iptables iptab
+		if !libcontainersystem.RunningInUserNS() {
+			return nil, fmt.Errorf("failed to set open file handler limit to 64000: %v", err)
+		}
+		klog.Errorf("failed to set open file handler limit to 64000: %v (running in UserNS)", err)
```
fixed

AkihiroSuda commented:
/retest

AkihiroSuda commented:
CI failure seems unrelated
```
I0716 01:45:32.884]   test/e2e/instrumentation/monitoring/metrics_grabber.go:84
I0716 01:45:32.884] STEP: Proxying to Pod through the API server
I0716 01:45:32.884] Jul 16 01:45:26.324: FAIL: Unexpected error:
I0716 01:45:32.884]     <*errors.errorString | 0xc002b708e0>: {
I0716 01:45:32.884]         s: "kube-controller-manager pod is not registered. Skipping ControllerManager's metrics gathering",
I0716 01:45:32.884]     }
I0716 01:45:32.885]     kube-controller-manager pod is not registered. Skipping ControllerManager's metrics gathering
I0716 01:45:32.885] occurred
skipped 159 lines unfold_more
I0716 01:45:32.950] Latency metrics for node e2e-b08efbf840-674b9-minion-group-xqj9
I0716 01:45:32.950] Jul 16 01:45:32.668: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
I0716 01:45:32.950] STEP: Destroying namespace "metrics-grabber-4134" for this suite.
I0716 01:45:32.950] 
I0716 01:45:32.950] 
I0716 01:45:32.950] • Failure [9.164 seconds]
I0716 01:45:32.951] [sig-instrumentation] MetricsGrabber
I0716 01:45:32.951] test/e2e/instrumentation/common/framework.go:23
I0716 01:45:32.951]   should grab all metrics from a ControllerManager. [It]
I0716 01:45:32.951]   test/e2e/instrumentation/monitoring/metrics_grabber.go:84
I0716 01:45:32.951] 
skipped 2607 lines unfold_more
I0716 01:46:24.351] Jul 16 01:46:20.788: INFO: Service reachability failing with error: error running /go/src/k8s.io/kubernetes/kubernetes/platforms/linux/amd64/kubectl --server=https://34.105.58.44 --kubeconfig=/workspace/.kube/config exec --namespace=services-2083 hairpin -- /bin/sh -x -c nc -zv -t -w 2 hairpin-test 8080:
```

liggitt commented:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92863/pull-kubernetes-node-e2e/1283624271305773056 is flake #93210

BenTheElder commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/92863/pull-kubernetes-verify/1287991226703810562
> Run ./hack/update-bazel.sh

AkihiroSuda commented:
Thanks @BenTheElder , updated

AkihiroSuda commented:
/test pull-kubernetes-integration

dashpole commented:
/hold
This needs to wait for the KEP to be merged as implementable.

AkihiroSuda commented:
@dashpole What do I need to do to get the [KEP](https://github.com/kubernetes/enhancements/pull/1371) merged?

dashpole commented:
@AkihiroSuda it needs approval from Dawn or Derek.  Bringing it up for discussion at sig-node is generally the best way to get  movement, or figure out what next steps are.

aojea commented:
@AkihiroSuda naive question, both kubelet and kube-proxy need to handle and modifgy iptables rules in the host, is this possible in rootless or are there any limitations?

AkihiroSuda commented:
> @AkihiroSuda naive question, both kubelet and kube-proxy need to handle and modifgy iptables rules in the host, is this possible in rootless or are there any limitations?

Possible, because we are in a user namespace + network namespace.
However, we will need a custom controller that exposes ports in the namespace to the host namespace, by calling RootlessKit API.

https://github.com/kubernetes/enhancements/pull/1371/files

> #### Network
>
>The node components need to be executed in a network namespace created along with the user namespace, because an unprivileged user cannot do privileged operations in the host network namespace.
As the components are executed inside a network namespace, `NodePorts` are not directly accessible from other hosts.
>
>An external controller should watch changes on `corev1.Service` resources and call [RootlessKit API](https://github.com/rootless-containers/rootlesskit/blob/v0.9.5/pkg/api/openapi.yaml) to set up port forwarding for the node ports.
>
>k3s implementation: https://github.com/rancher/k3s/blob/v1.17.2+k3s1/pkg/rootlessports/controller.go#L92-L96

aojea commented:
> As the components are executed inside a network namespace, NodePorts are not directly accessible from other hosts.

:thinking: , is only nodeports affected? 
From the top of my head I'm not sure that this will work in  multinode, pods need to be able to access services and pods in other nodes, and that requires packet mangling ... but I need more info, can you point me to some place with more details about the network details in rootless?

AkihiroSuda commented:
Aside from NodePorts, just exposing 10250/tcp for Kubelet and 8472/udp for VXLAN (Flannel) should be fine.

Pods can communicate with pods in other nodes via VXLAN.

The VXLAN works inside the namespace, and does not affect the host namespace, and does not require the host root.

We have a Docker Compose stack for testing multi-node cluster: https://github.com/rootless-containers/usernetes/blob/master/docker-compose.yml

AndreMouche commented:
/lgtm

AkihiroSuda commented:
Simplified the PR.

The latest revision of this PR just modifies kubelet and kube-proxy to relax sysctl errors and `RLIMIT_NOFILE` errors.

The list of relaxed sysctl keys in this PR:
- `vm.overcommit_memory` (kubelet)
- `vm.panic_on_oom` (kubelet)
- `kernel.panic` (kubelet)
- `kernel.panic_on_oops` (kubelet)
- `kernel.keys.root_maxkeys` (kubelet)
- `kernel.keys.root_maxbytes` (kubelet)
- `net.netfilter.nf_conntrack_max` (kube-proxy)
- `net.netfilter.nf_conntrack_tcp_timeout_established` (kube-proxy)
- `net.netfilter.nf_conntrack_tcp_timeout_close_wait` (kube-proxy)

AkihiroSuda commented:
Maintainers: could you remove the following labels from this PR: `kind/api-change` `sig/network` `sig/testing` `area/test`

dims commented:
/milestone v1.21

aojea commented:
it seems that the KEP is still blocked https://github.com/kubernetes/enhancements/pull/1371 :/

AkihiroSuda commented:
Reduced the amount of the patches:
```console
$ git diff upstream/master --stat
 pkg/kubelet/cm/container_manager_linux.go | 7 ++++++-
 pkg/proxy/userspace/proxier.go            | 6 +++++-
 2 files changed, 11 insertions(+), 2 deletions(-)
```

erismaster commented:
Hello! I am from the Bug Triage team! Friendly reminder that code freeze is starting March 9th, 2021 (about 1 week from now). We want to ensure that each PR has a chance to be merged before code freeze.
As the PR is tagged for 1.21, is it still planned for this release?

dims commented:
@SergeyKanzhelev @ehashman can you please look? this is been going on for a while for what seems like a simple enough patch.

erismaster commented:
Hi @AkihiroSuda, we are past code freeze for 1.21, I am going to move this to the 1.22 milestone. 

/milestone 1.22

erismaster commented:
/milestone v1.22

AkihiroSuda commented:
The KEP https://github.com/kubernetes/enhancements/pull/1371 was merged today. I'll update this PR soon to reflect the latest KEP.

AkihiroSuda committed:
New feature gate: KubeletInUserNamespace

Enables support for running kubelet in a user namespace.
The user namespace has to be created before running kubelet.
All the node components such as CRI need to be running in the same user namespace.

See kubernetes/enhancements PR 1371 (merged) and issue 2033.

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
kubelet: ignore /dev/kmsg error when running in userns

oomwatcher.NewWatcher returns "open /dev/kmsg: operation not permitted" error,
when running with sysctl value `kernel.dmesg_restrict=1`.

The error is negligible for KubeletInUserNamespace.

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda commented:
Updated the PR

dims commented:
/assign @SergeyKanzhelev @ehashman

aojea commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd
metricsgrabber flake

aojea commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -454,6 +455,13 @@ func setupKernelTunables(option KernelTunableBehavior) error {
 			if err != nil {
+				if libcontainersystem.RunningInUserNS() {
+					if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.KubeletInUserNamespace) {
+						klog.Warningf("Updating kernel flag failed: %v: %v (running in UserNS, ignoring)", flag, err)
```
I think that they've already migrated this to the new logging format 
```
				klog.ErrorS(err, "Failed to get container name for process")
```

AkihiroSuda replied:
updated, thanks

aojea commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -454,6 +455,13 @@ func setupKernelTunables(option KernelTunableBehavior) error {
+						klog.Warningf("Updating kernel flag failed: %v: %v (running in UserNS, ignoring)", flag, err)
+						continue
+					}
+					klog.Errorf("Updating kernel flag failed: %v: %v (Hint: enable KubeletInUserNamespace feature flag to ignore the error)", flag, err)
```
klog.Errors

AkihiroSuda replied:
updated

aojea commented:
LGTM for kube-proxy
This is more a sig-node thing :)

ehashman commented:
/priority important-soon

AkihiroSuda commented:
/retest

ehashman commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -454,6 +455,13 @@ func setupKernelTunables(option KernelTunableBehavior) error {
 			klog.V(2).InfoS("Updating kernel flag", "flag", flag, "expectedValue", expectedValue, "actualValue", val)
 			err = sysctl.SetSysctl(flag, expectedValue)
 			if err != nil {
+				if libcontainersystem.RunningInUserNS() {
+					if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.KubeletInUserNamespace) {
```
These can be combined into one condition, and the feature gate check should come first.

AkihiroSuda replied:
updated

ehashman commented on pkg/kubelet/kubelet.go:
```diff
@@ -479,7 +480,17 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 	oomWatcher, err := oomwatcher.NewWatcher(kubeDeps.Recorder)
 	if err != nil {
-		return nil, err
+		if !libcontainersystem.RunningInUserNS() {
+			return nil, err
+		}
+		// oomwatcher.NewWatcher returns "open /dev/kmsg: operation not permitted" error,
+		// when running with sysctl value `kernel.dmesg_restrict=1`.
+		if !utilfeature.DefaultFeatureGate.Enabled(features.KubeletInUserNamespace) {
+			klog.ErrorS(err, "Failed to create an oomWatcher (running in UserNS, Hint: enable KubeletInUserNamespace feature flag to ignore the error)")
+			return nil, err
+		}
+		klog.V(2).InfoS("Failed to create an oomWatcher (running in UserNS, ignoring)", "err", err)
+		oomWatcher = nil
```
This whole change should go behind a feature flag.

AkihiroSuda replied:
updated

ehashman commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -231,7 +232,10 @@ func NewCustomProxier(loadBalancer LoadBalancer, listenIP net.IP, iptables iptab
 	err = setRLimit(64 * 1000)
 	if err != nil {
-		return nil, fmt.Errorf("failed to set open file handler limit: %v", err)
+		if !libcontainersystem.RunningInUserNS() {
+			return nil, fmt.Errorf("failed to set open file handler limit to 64000: %v", err)
+		}
```
This needs a feature gate.

AkihiroSuda replied:
updated

h-vetinari commented:
> @ehashman moved this from Needs Reviewer to Waiting on Author in [SIG Node PR Triage](https://github.com/orgs/kubernetes/projects/49#card-55620039) 4 days ago

I'm almost certain this isn't waiting on author anymore

ehashman commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -231,7 +234,11 @@ func NewCustomProxier(loadBalancer LoadBalancer, listenIP net.IP, iptables iptab
 	if err != nil {
-		return nil, fmt.Errorf("failed to set open file handler limit: %v", err)
+		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.KubeletInUserNamespace) && libcontainersystem.RunningInUserNS() {
+			klog.ErrorS(err, "failed to set open file handler limit to 64000 (running in UserNS, ignoring the error)")
```
```suggestion
			klog.ErrorS(err, "failed to set open file handler limit to 64000 (running in UserNS, ignoring)")
```

for consistency

AkihiroSuda replied:
fixed

ehashman reviewed (commented):
/lgtm

ehashman commented:
/hold cancel

BenTheElder commented:
/retest

ehashman commented:
re-
/lgtm

SergeyKanzhelev commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -454,6 +455,11 @@ func setupKernelTunables(option KernelTunableBehavior) error {
+					klog.V(2).InfoS("Updating kernel flag failed (running in UserNS, ignoring)", "flag", flag, "err", err)
+					continue
+				}
+				klog.ErrorS(err, "Updating kernel flag failed (Hint: enable KubeletInUserNamespace feature flag to ignore the error if running in a user namespace)", "flag", flag)
```
You can check whether running in user namespace, so no need to output hint in case kubelet running outside of it, right?

AkihiroSuda replied:
updated

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -479,7 +480,19 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
+				klog.V(2).InfoS("Failed to create an oomWatcher (running in UserNS, ignoring)", "err", err)
+				oomWatcher = nil
+			} else {
+				klog.ErrorS(err, "Failed to create an oomWatcher (running in UserNS, Hint: enable KubeletInUserNamespace feature flag to ignore the error)")
```
hint is misleading as the error is printed outside the check `libcontainersystem.RunningInUserNS()`.

AkihiroSuda replied:
updated

SergeyKanzhelev commented:
/lgtm cancel

to improve error hints

dims commented:
/approve
/lgtm

thanks for updating after comments @AkihiroSuda

AkihiroSuda commented:
/assign @andrewsykim @derekwaynecarr

SergeyKanzhelev commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -454,6 +455,13 @@ func setupKernelTunables(option KernelTunableBehavior) error {
+						klog.V(2).InfoS("Updating kernel flag failed (running in UserNS, ignoring)", "flag", flag, "err", err)
+						continue
+					}
+					klog.ErrorS(err, "Updating kernel flag failed (Hint: enable KubeletInUserNamespace feature flag to ignore the error)", "flag", flag)
```
this case should add error to the errors list
```
errList = append(errList, err)
```


?

/hold

AkihiroSuda replied:
Yes, `errList = append(errList, err)` is called in L465

SergeyKanzhelev replied:
I see, I was confused by continue `above`, sorry

ehashman commented:
@SergeyKanzhelev can you PTAL and see if your hold is addressed?

kolyshkin commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -33,6 +33,7 @@ import (
 	cgroupfs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
 	cgroupfs2 "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
 	"github.com/opencontainers/runc/libcontainer/configs"
+	libcontainersystem "github.com/opencontainers/runc/libcontainer/system"
```
This should be `"github.com/opencontainers/runc/libcontainer/userns` because of https://github.com/opencontainers/runc/pull/2850 (which is part of runc v1.0.0-rc94).

AkihiroSuda replied:
updated

kolyshkin commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -26,14 +26,17 @@ import (
 	"sync/atomic"
 	"time"
 
+	libcontainersystem "github.com/opencontainers/runc/libcontainer/system"
```
ditto, this should be `"github.com/opencontainers/runc/libcontainer/userns"`

AkihiroSuda replied:
updated

kolyshkin commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -231,7 +234,11 @@ func NewCustomProxier(loadBalancer LoadBalancer, listenIP net.IP, iptables iptab
+		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.KubeletInUserNamespace) && libcontainersystem.RunningInUserNS() {
+			klog.ErrorS(err, "failed to set open file handler limit to 64000 (running in UserNS, ignoring)")
+		} else {
+			return nil, fmt.Errorf("failed to set open file handler limit to 64000: %v", err)
```
I think the new code should use `%w` for error wrapping.

kolyshkin replied:
(this is not a new code, but maybe makes sense to fix it while moving it)

AkihiroSuda replied:
updated

kolyshkin commented on pkg/kubelet/kubelet.go:
```diff
@@ -37,6 +37,7 @@ import (
 	"k8s.io/mount-utils"
 	"k8s.io/utils/integer"
 
+	libcontainersystem "github.com/opencontainers/runc/libcontainer/system"
```
ditto, this should be `"github.com/opencontainers/runc/libcontainer/userns"`

AkihiroSuda replied:
updated

kolyshkin commented on pkg/kubelet/kubelet.go:
```diff
@@ -1356,8 +1369,10 @@ func (kl *Kubelet) initializeModules() error {
-		return fmt.Errorf("failed to start OOM watcher %v", err)
+	if kl.oomWatcher != nil {
+		if err := kl.oomWatcher.Start(kl.nodeRef); err != nil {
+			return fmt.Errorf("failed to start OOM watcher %v", err)
```
Use `%w` for error wrapping?

kolyshkin replied:
...and a comma to separate the message from the error, i.e.
```go
return fmt.Errorf("failed to start OOM watcher: %w", err)
```

(I understand it's not your code and you're just moving it around, but while at it it makes sense to fix these nitpicks.

AkihiroSuda replied:
updated

kolyshkin reviewed (changes requested):
The import path should be fixed to "github.com/opencontainers/runc/libcontainer/userns", plus some minor nits.

AkihiroSuda commented:
@kolyshkin Thanks, updated

kolyshkin reviewed (approved):
/lgtm (non-binding)

AkihiroSuda commented:
Rebased

AkihiroSuda commented:
Rerebased

AkihiroSuda commented:
/retest

AkihiroSuda commented:
/retest

cbandy commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -231,7 +234,11 @@ func NewCustomProxier(loadBalancer LoadBalancer, listenIP net.IP, iptables iptab
 	if err != nil {
-		return nil, fmt.Errorf("failed to set open file handler limit: %v", err)
+		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.KubeletInUserNamespace) && libcontaineruserns.RunningInUserNS() {
+			klog.ErrorS(err, "failed to set open file handler limit to 64000 (running in UserNS, ignoring)")
```
In the other files, `(running in UserNS, ignoring)` is emitted using `V(2).InfoS`. Is `ErrorS` intentional here?

AkihiroSuda replied:
fixed

AkihiroSuda commented:
/retest

AkihiroSuda commented:
/retest

AkihiroSuda commented:
rebased

AkihiroSuda commented:
Rebased.

Can we get this merged soon?

AkihiroSuda commented:
/retest

AkihiroSuda commented:
/retest

warmchang commented:
/test pull-kubernetes-node-e2e
/test pull-kubernetes-e2e-gce-alpha-features

BenTheElder commented:
it will not merge because it still lacks approval https://github.com/kubernetes/kubernetes/pull/92863#issuecomment-867967971 

> Needs approval from an approver in each of these files:
> ~pkg/features/OWNERS~ [dims]
> pkg/kubelet/OWNERS
> pkg/proxy/userspace/OWNERS

that would be @derekwaynecarr + @andrewsykim as currently assigned for kubelet and proxy, but if they're not available:

https://github.com/kubernetes/kubernetes/blob/ba008d6131b06ad9a474c69681de0712c470352e/pkg/kubelet/OWNERS#L4
https://github.com/kubernetes/kubernetes/blob/ba008d6131b06ad9a474c69681de0712c470352e/OWNERS_ALIASES#L268

though also for review prior to approval we're still waiting on @SergeyKanzhelev https://github.com/kubernetes/kubernetes/pull/92863#issuecomment-866414850

BenTheElder commented:
after this merges the alpha features job will probably need updating to not enable _this_ alpha feature ...
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/92863/pull-kubernetes-e2e-gce-alpha-features/1411008443313033216/

AkihiroSuda commented:
@derekwaynecarr @mrunalp (SIG-node) @andrewsykim @aojea (SIG-network) @SergeyKanzhelev
Could you take a look?

/retest

AkihiroSuda commented:
<details>

<summary>CI failure is unrelated </summary>

<p>

```
I0705 07:27:49.411] NAME                 STATUS      MESSAGE                                                                                       ERROR
I0705 07:27:49.412] scheduler            Unhealthy   Get "http://127.0.0.1:10251/healthz": dial tcp 127.0.0.1:10251: connect: connection refused   
I0705 07:27:49.412] etcd-0               Healthy     {"health":"true","reason":""}                                                                 
I0705 07:27:49.412] etcd-1               Healthy     {"health":"true","reason":""}                                                                 
I0705 07:27:49.413] controller-manager   Healthy     ok                                                                                            
I0705 07:27:49.415] Validation returned one or more failed components. Cluster is probably broken.
```

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/92863/pull-kubernetes-e2e-gce-alpha-features/1411942353094053888/
</p>

</details>

/retest

AkihiroSuda commented:
/retest

aojea commented:
/approve
sig-network changes

neolit123 commented:
> Added support for rootless mode

i recommend amending the release note to be more specific of what the change is about and what components are being modified.

AkihiroSuda commented:
@neolit123 
```release-note
Adds feature gate `KubeletInUserNamespace` which enables support for running kubelet in a user namespace.

The user namespace has to be created before running kubelet.
All the node components such as CRI need to be running in the same user namespace.

When the feature gate is enabled, kubelet ignores errors that happens during setting the following sysctl values: `vm.overcommit_memory`, `vm.panic_on_oom`, `kernel.panic`, `kernel.panic_on_oops`, `kernel.keys.root_maxkeys`, `kernel.keys.root_maxbytes`. (These sysctl values for the host, not for the containers)

kubelet also ignores an error during opening `/dev/kmsg`.
This feature gate also allows kube-proxy to ignore an error during setting `RLIMIT_NOFILE`.

This feature gate is especially useful for running Kubernetes inside Rootless Docker/Podman with `kind` or `minikube`.
```

/retest

AkihiroSuda commented:
/retest

AkihiroSuda commented:
Rebased

AkihiroSuda commented:
@kubernetes/sig-node-pr-reviews Can we get this reviewed before v1.22 Code Freeze on July 8th? 🙏

dims commented:
/skip

dims commented:
/hold cancel

the KEP is ready for 1.22

ehashman reviewed (commented):
/lgtm

AkihiroSuda commented:
Thanks all for reviewing and approval!

Opened a doc PR: https://github.com/kubernetes/website/pull/28827
